src/detect-engine-alert.c

```diff
@@ -209,6 +208,7 @@ int PacketAlertAppend(DetectEngineThreadCtx *det_ctx, const Signature *s,
         const uint16_t target_pos = i + 1;
         const uint16_t space_post_target = packet_alert_max - 1 - i;
         const uint16_t to_move = MIN(space_post_target, (p->alerts.cnt - i));
+        BUG_ON(to_move == 0);
```

Comment:
It was indeed triggering before, but after the line 193 patch, it has stopped, to me. 

> It can be 0 : if space_post_target is 0 which can be the case if you have like packet_alert_max= 3, you have 3 alerts with num 3,5,7 and you want to add one with num 6

Would a test like this be a good addition, to check for this type of situation? It passes, for me.
https://github.com/jufajardini/suricata-verify/commit/1a5688d26bff0fb775c02a3f8a64a36609713824

Reply:
> but after the line 193 patch, it has stopped, to me.

You mean the patch at line 186, right ?
I think this check is wrong
```
if (p->alerts.cnt == packet_alert_max)
        return 0;
```
in the sense, that if it prevents the `BUG_ON` which is good, we do not get the expected outcome.

That is if `packet_alert_max == 3` and signatures 3,5,7,6 are added as alerts, in the end, we get 3,5,7 instead of 3,5,6

> Would a test like this be a good addition, to check for this type of situation? It passes, for me.

I do not know enough about the context to tell if the S-V test is good or not : I do not know how Suricata will add the alerts, like 3,5,7,6 or 3,5,6,7...
So, there should be more a unit test...
Od do you know more about the context ? ie the order of the calls to `PacketAlertAppend`

Reply:
I was talking this -> `if (p->alerts.cnt == 0 || p->alerts.alerts[p->alerts.cnt - 1].num <= s->num)`
It used to be just `p->alerts.alerts[p->alerts.cnt - 1].num < s->num` and that could trigger the `BUG_ON`, but once I changed that, I didn't see it be triggered again.

Reply:
I'll try to add a unittest, too

Reply:
> > but after the line 193 patch, it has stopped, to me.
> 
> You mean the patch at line 186, right ? I think this check is wrong
> 
> ```
> if (p->alerts.cnt == packet_alert_max)
>         return 0;
> ```
> 
> in the sense, that if it prevents the `BUG_ON` which is good, we do not get the expected outcome.
> 
> That is if `packet_alert_max == 3` and signatures 3,5,7,6 are added as alerts, in the end, we get 3,5,7 instead of 3,5,6

So the expected behavior if we have more sig matches than packet_alert_ max is to drop some older alert for the new? Is that so because once we have added an alert to the queue, it is already logged/ taken care of, so we can let go of it?

> 
> > Would a test like this be a good addition, to check for this type of situation? It passes, for me.
> 
> I do not know enough about the context to tell if the S-V test is good or not : I do not know how Suricata will add the alerts, like 3,5,7,6 or 3,5,6,7... So, there should be more a unit test... Od do you know more about the context ? ie the order of the calls to `PacketAlertAppend`

When I run that s-v test with gdb, the result is that signatures are appended 3,5,6,7, which means we do not append sig 7

Reply:
I'm thinking that with https://redmine.openinfosecfoundation.org/issues/4943 the behavior for `PacketAlertAppend` is to just append the signatures that Suri has decided should stay...

Reply:
> So the expected behavior if we have more sig matches than packet_alert_ max is to drop some older alert for the new? Is that so because once we have added an alert to the queue, it is already logged/ taken care of, so we can let go of it?

There is a priority order on the signatures (which is not exactly their signature id I think)
So, we should keep the highest priority ones

Reply:
I see. This logic seem to be done elsewhere, at the moment...